Fix #6261: New pragma `--trace-imports`

knisht opened the pull request:
This pragma displays notification when the type-checking of a module is finished.

I welcome any suggestions on making the names clearer :).

Fixes #6261.

knisht committed:
[re agda#6261] Introduce the pragma `--trace-imports`

This pragma displays notification when the type-checking of a module is finished.

nad commented:
Thanks!

I don't think this should be a pragma option. The user of a library should be able to control this, but I don't think the developer of a library should be able to force users to see this kind of output.

Currently one can turn off the `Checking` messages using `-v0`. Is that still possible with these changes?

knisht committed:
[re agda#6261] Convert the pragma to a command-line option

knisht commented:
Thanks for the feedback.

My implementation is a bit tricky: if the user starts the type-checking from some file in emacs, then the `--trace-imports` pragma of this file becomes applied to all loaded files. This way, the developer of a library does not have any way to propagate this pragma to the user.
But I get your point. Honestly, I also don't see any use of this option as a pragma, I implemented it that way because why not, so I'll leave it as a plain CLI option.

Yes, it is possible. I have added a test for it.

andreasabel commented:
Discussion with Konstantin 2022-11-28: use verbosity levels (1,2,3) instead of names "on-enter",  "on-exit", "all".

nad commented:
> Discussion with Konstantin 2022-11-28: use verbosity levels (1,2,3) instead of names "on-enter", "on-exit", "all".

Why?

knisht commented:
For a person who doesn't know what is this option about, the name "on-exit" seems opposite to "on-enter", while in reality it is augmenting "on-enter". Also, there is a mode "all", which does not clearly show that it is going to display information about loaded interfaces. We thought about binding the names to the type of module they access, like "enter-uncompiled", "enter-exit-uncompiled", but these variants are awfully long. 
So in the end we decided to use numbers. They successfully represent the cumulativity of the modes (an analogy with verbosity levels), and since they are impossible to interpret without knowing the logic behind them, the user will be forced to read the documentation on the option. 

I still welcome any ideas on better naming.

knisht committed:
[re agda#6333] Use levels instead of textual options

andreasabel commented:
Rendering of the docs here: https://agda--6347.org.readthedocs.build/en/6347/tools/command-line-options.html#cmdoption-trace-imports

andreasabel reviewed (approved):
Thanks, good work!

andreasabel committed:
Merge branch 'master' into 6261

andreasabel committed:
Merge branch 'master' into knisht-6261

andreasabel commented on test/interaction/Issue6261.out:
```diff
@@ -0,0 +1,36 @@
+(agda2-info-action "*Type-checking*" "" nil)
+(agda2-highlight-clear)
+(agda2-info-action "*Type-checking*" "Loading Issue6261 (_build/2.6.4/agda/Issue6261.agdai). " t)
+(agda2-info-action "*Type-checking*" " Loading Issue6261a (_build/2.6.4/agda/Issue6261a.agdai). " t)
```
This message apparently can come later, there is a race condition on CI:
https://github.com/agda/agda/actions/runs/5765746926/job/15632876653#step:11:201 reports the following diff:
```
 (agda2-info-action "*Type-checking*" "Loading Issue6261 (_build/2.6.4/agda/Issue6261.agdai). " t)
-(agda2-info-action "*Type-checking*" " Loading Issue6261a (_build/2.6.4/agda/Issue6261a.agdai). " t)
 (agda2-info-action "*Type-checking*" " Loading Issue6261b (_build/2.6.4/agda/Issue6261b.agdai). " t)
 (agda2-info-action "*Type-checking*" " Loading Issue6261c (_build/2.6.4/agda/Issue6261c.agdai). " t)
+(agda2-info-action "*Type-checking*" " Loading Issue6261a (_build/2.6.4/agda/Issue6261a.agdai). " t)
```
